### PR TITLE
Add some more WORKFLOW_DETECTORS_... options

### PR DESCRIPTION
**Pull request by davidrohr**
@shahor02 : This adds `WORKFLOW_DETECTORS_CALIB`, `WORKFLOW_DETECTORS_RECO`, `WORKFLOW_DETECTORS_CTF`.
I tested several combinations, but no guarantee for completeness.

I thought it is a good idea to add also the CTF variable (despite actually wanting less variables). The problem is that otherwise you have problems e.g. with TPC and TOF. TPC CTF needs the GPU reco, so either the CTF option must implicitly enable the GPU reco, or the `WORKFLOW_DETECTORS_RECO!=TPC` must remove the TPC from the CTF. With both DETECTOR variables, I can do a logical OR, so if you want to get rid of the TPC completely, remove TPC from both of them. If you remove it only from `_RECO` but not from `_CTF` the part that is required for the CTF will still run.

I decided not to add all these fields to the AliECS GUI, since I think it should not be needed for normal operations by the shifters, and I don't want to overload the GUI. Instead the AliECS gui has a field `EXTRA ENVIRONMENT VARIABLES`. When it is deployed, one can put variables in shell syntax there, so it would be `EXTRA_ENV_VARS="WORKFLOW_DETECTORS_CALIB=TPC WORKFLOW_DETECTORS_RECO=TPC"`. This gives us all flexibility for experts, while keeping the default interface simple, and it is also transparent in case we add more options.

I'll also add one more option `WORKFLOW_DETECTORS_MATCHING` son, which will be able to enable/disable all matchings and vertexing. But currently still thinking about how to do it in the cleanest way, thus I am opening this PR already as is.

### CODE REVIEW

**Comment by davidrohr** [comment]
Now added also `WORKFLOW_DETECTORS_MATCHING`. First I tried to include some magic, e.g. to disable ITS-TPC-TRD matching when ITS-TPC matching is disabled, and depending on the detector presence. But this only complicates the workflow script. Thus, since this is a debug option for experts anyway, one can either leave it empty for the default matching for sync and async mode respectively, or give an explicit list of matching vertexing algorithms to run, or use `ALL` to enable all of them.

File: production/dpl-workflow.sh
Comment (shahor02) [inline production/dpl-workflow.sh]: this loop with 1 input per detector will not work for all detectors except TOF, perhaps we should add a function which can be called from multiple times in the single loop, something like

```
add_to_proxy_input()
{
  if (( $# < 2 )); then
    echo "destination and at least 1 input to add is expected"
    exit 1
  fi
  for ((i = 2; i <= $#; i++ )); do
    count=`grep -o ':' <<< "${!1}"  | wc -l
    $1+=";INP${count}:{i}"
  }
}
```
and use it as 

```
  for i in `echo "$WORKFLOW_DETECTORS" | sed "s/,/ /g"`; do
    if has_detector_flp_processing $i; then
      case $i in
        TOF)
          add_to_proxy_input PROXY_INSPEC "$i/CRAWDATA" ;;
        FT0 | FV0 | FDD )
          add_to_proxy_input PROXY_INSPEC "$i/DIGITSBC"  "$i/DIGITSCH" ;;
        PHS)
          add_to_proxy_input PROXY_INSPEC "$i/CELLS"  "$i/CELLTRIGREC" ;;
        CPV)
          add_to_proxy_input PROXY_INSPEC "$i/DIGITS"  "$i/DIGITTRIGREC" "$i/RAWHWERRORS" ;;
        EMC)
          add_to_proxy_input PROXY_INSPEC "$i/CELLS"  "$i/CELLSTRGR" "$i/DECODERERR" ;;
        *)
          echo Input type for detector $i with FLP processing not defined 1>&2
          exit 1;;
      esac
    else
      add_to_proxy_input PROXY_INSPEC "$i/RAWDATA"
    fi
  done
```

For the QC, we have 2 additions:
```
[ -z "$QC_JSON_FT0" ] && QC_JSON_FT0=/home/afurs/work/epn/configs/qc/ft0-digits-qc-ds.json
[ -z "$QC_JSON_FV0" ] && QC_JSON_FV0=/home/afurs/work/epn/configs/qc/fv0-digits-qc-ds.json
```

**Review by shahor02 — COMMENTED** [review]
Hi @davidrohr

Superficially looks fine except for one point below. I think eventually we would need to go for the detector-set dependent matching workflows. It complicates the things, but hacking the dpl-workfow each time we need to generate non-standard configuration would complicate the bookkeeping and will be more error-prone...

**Comment by davidrohr** [comment]
Well, I am allready checking whether the detectors are in the Workflow_detectors_reco and enable the matching specs only then. Isn't this sufficient?

The matching option is only an option on top, i.e. both all detectors must be present, and the respective matching must be activated. Or do I miss something?

Kind Regards
David Rohr

Sent from my mobile. (Excuse the typos!)

File: production/dpl-workflow.sh
Comment (shahor02) [inline production/dpl-workflow.sh]: There was a typo here, should be `FV0`

File: production/dpl-workflow.sh
Comment (shahor02) [inline production/dpl-workflow.sh]: I think we also need to compose incrementally the input sources for P/S vertexing (in principle, can be the same, collecting all available track).
I would not forbid the vertexing if all {ITS TPC TRD TOF FT0 MCH MFT} are not there. Technically, ITS alone is enough  (and necessary) to run primary vertexing,  for the secondary one it is ITS+TPC (and soon MFT will join for forward secondaries).

**Review by shahor02 — COMMENTED** [review]
Hi @davidrohr 

After having closer look I think the matching part is fine, sorry for the noise.
But the vertexers will require additional treatment + there was an old typo, see below.

File: production/dpl-workflow.sh
Comment (davidrohr) [reply production/dpl-workflow.sh]: Sure, but since the vertexers were for async only I didn't spend much time so far. Do the vertexers have a `--track-sources` option or similar?

File: production/dpl-workflow.sh
Comment (davidrohr) [reply production/dpl-workflow.sh]: thx, indeed, don't have much time now, but will include your suggestions later today and push a new version.

File: production/dpl-workflow.sh
Comment (shahor02) [reply production/dpl-workflow.sh]: `o2-primary-vertexing-workflow` has 2 input sources lists:
`--vertexing-sources <used_for_PVertex_finding>`
`--vetex-track-matching-sources <sources to be associated with vertices matching time-wise>`

`o2-secondary-vertexing-workflow` has simply 
`--vertexing-sources <sources used for v0 and cascade finding>`

**Comment by davidrohr** [comment]
@shahor02 : Should be everything in now. Before merging I want to run it also through the full system test, just waiting for my O2 compilation to finish.

File: production/dpl-workflow.sh
Comment (shahor02) [inline production/dpl-workflow.sh:270]: Better not use subspecs. E.g. EMC gets data from 2 FLPs, so the subspec will likely be the FEEID of the 1st link (they have open PR on hold for merging of such data).

**Review by shahor02 — COMMENTED** [review]
thanks, looks nice, just 1 comment below

File: production/dpl-workflow.sh
Comment (shahor02) [reply production/dpl-workflow.sh:270]: Actually, at least for FT0 | FV0 | FDD and CPV the subspec 0 is needed for sure, since downstream devices expect this and there is anyway 1 FLP. For the EMC it is certainly not needed as 2 streams need to be merged. Will need to clarify for the PHOS.

**Comment by davidrohr** [comment]
I have tried without the subspec, but it didn't work. I added the subspecs only where it was necessary. It seems the consumers downstream request this subspec.

**Comment by davidrohr** [comment]
@shahor02 : so currently it breaks the FST, I'll fix that later.
Afterwards I'd merge it in case you have no additional comments.
I'd remove the subspecs once O2 is fixed such that they are no longer requested.

**Review by shahor02 — APPROVED** [review]

**Pull request merged**
